### DISCUSSION TITLE
k3s agent node has an alerting from prometheus: `Client certificate is about to expire. A client certificate used to authenticate to the apiserver is expiring in less than 7.0 days.`

### DISCUSSION THREAD

**Discussion by wang-xiaowu** [k3s-io/k3s#8227] [Q&A] [answered]
recently i receive the alerting from prometheus: 
```
summary: Client certificate is about to expire. 
description: A client certificate used to authenticate to the apiserver is expiring in less than 7.0 days.
```

it from the agent node
and it exactly has less than 7 days
```bash
[root@k3s-prod-node1 ~]# openssl x509 -noout -in /var/lib/rancher/k3s/agent/client-k3s-controller.crt -enddate
notAfter=Aug 30 02:37:05 2023 GMT
```
but when i run the command `k3s certificate rotate`, it shows error `FATA[0001] stat /var/lib/rancher/k3s/server/tls: no such file or directory`. finnaly i restart it

but like this [issue](https://github.com/k3s-io/k3s/issues/5111) said, agent needn't to rotate the certs, right?
so does this alerting has any effect?

**Comment by brandond** [comment, marked as answer]
The kubelet certificates are reissued whenever the agent is started. Just restart the agent at least once a year and you'll be fine.

You're patching your nodes at least once a year, right?

**Reply by wang-xiaowu** [reply]
eh，no
i thought agent needn't to be restart
and when i starting to use k3s, its version is 1.21, there is no `k3s certificate rotate`
so i write the crontab shell only run on the server nodes
```bash
#!/usr/bin/env bash
tempCertEndDate=$(openssl x509 -noout -in /var/lib/rancher/k3s/server/tls/client-admin.crt -enddate)
certEndDate=$(echo "$tempCertEndDate" |cut -d"=" -f2)
tempFormatEndDate=$(date -u -d "$certEndDate" +'%Y-%m-%d %H:%M:%S')
formatEndDate=$(date +%s -d "$tempFormatEndDate")
now=$(date +%s)
restartDate=$(date +"%Y-%m-%d %H:%M:%S")
days=$((($formatEndDate - $now)/86400))
if [ $days -lt 90 ]; then
    mkdir -p /var/log/k3s
    systemctl restart k3s
    echo -e $(date +"%Y-%m-%d %H:%M:%S")" - 证书即将过期，重启k3s服务，剩余天数："$days >> /var/log/k3s/restart_k3s.log
fi
```

**Reply by wang-xiaowu** [reply]
plus：now i am using v1.24.16+k3s1

**Reply by brandond** [reply]
> no i thought agent needn't to be restart

That is incorrect.

Agents need to be restarted if you want them to get new client certs. Servers need to be restarted after running the cert rotation command, or within 90 days of server cert expiration to trigger automatic extension.

**Reply by wang-xiaowu** [reply]
i see
thx
